src/libraries/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/SignerInfo.cs

```diff
@@ -591,100 +587,238 @@ private bool CheckHash(bool compatMode)
+            {
+                foreach (AttributeAsn attr in signedAttributes)
+                {
+                    attr.Encode(writer);
 
-                            var digestAttr = (Pkcs9MessageDigest)obj.Values[0];
+                    // .NET Framework doesn't seem to validate the content type attribute,
+                    // so we won't, either.
 
-                            if (!contentDigest.AsSpan().SequenceEqual(digestAttr.MessageDigest))
-                            {
-                                throw new CryptographicException(SR.Cryptography_BadHashValue);
-                            }
+                    if (attr.AttrType == Oids.MessageDigest)
+                    {
+                        CryptographicAttributeObject obj = PkcsHelpers.MakeAttribute(attr);
 
-                            invalid = false;
+                        if (obj.Values.Count != 1)
+                        {
+                            throw new CryptographicException(SR.Cryptography_BadHashValue);
                         }
-                    }
 
-                    if (compatMode)
-                    {
-                        writer.PopSequence();
+                        var digestAttr = (Pkcs9MessageDigest)obj.Values[0];
+
+                        if (!digest.SequenceEqual(digestAttr.MessageDigest))
+                        {
+                            throw new CryptographicException(SR.Cryptography_BadHashValue);
+                        }
 
-                        byte[] encoded = writer.Encode();
-                        encoded[0] = 0x31;
-                        hasher.AppendData(encoded);
+                        hasMatchingDigestAttr = true;
                     }
-                    else
+                    else if (attr.AttrType == Oids.ContentType)
                     {
-                        writer.PopSetOf();
+                        hasContentAttr = true;
+                    }
+                }
+            }
 
+            // Message-digest is required when signed attributes are present.
+            if (!hasMatchingDigestAttr ||
+                needsContentAttr && !hasContentAttr)
+            {
+                throw new CryptographicException(SR.Cryptography_Cms_MissingAuthenticatedAttribute);
+            }
+
+            if (compatMode)
+            {
+                byte[] encoded = writer.Encode();
+                encoded[0] = 0x31;
+                return verify(encoded);
+            }
+            else
+            {
 #if NET9_0_OR_GREATER
-                        writer.Encode(hasher, static (hasher, encoded) => hasher.AppendData(encoded));
+                return writer.Encode(verify, static (verify, encoded) => verify(encoded));
+#else
+                return verify(writer.Encode());
+#endif
+            }
+        }
+
+        private delegate bool VerifyCallback(ReadOnlySpan<byte> contentToVerify);
+
+        private bool VerifyHashed(bool compatMode, VerifyCallback verify)
```

Comment:
Another place where "Hashed" feels weird.  "VerifyHashed" to me sounds like "Verify it was hashed", not "Verify the signature for algorithms that utilize a pre-hashing step".  HashedVerify does a better job of the latter, to my ear.